cmds/tui.c

```diff
@@ -2079,6 +2087,170 @@ static void tui_window_move_end(struct tui_window *win)
 	}
 }
 
+static void tui_window_move_scroll_line_up(struct tui_window *win)
+{
+	void *node = win->ops->next(win, win->top, true);
+
+	/* cannot go top next */
+	if (node == NULL)
+		return;
+
+	/* can go top next */
+	win->top_index++;
+	if (win->ops->needs_blank(win, win->top, node))
+		win->top_index++;
+
+	/* update win->top */
+	win->top = node;
+
+	/* update win->curr - out of cursor line case*/
+	if (win->top_index > win->curr_index) {
+		win->curr = win->top;
+		win->curr_index = win->top_index;
+	}
+}
+
+static void tui_window_move_scroll_line_down(struct tui_window *win)
+{
+	void *node = win->ops->prev(win, win->top, true);
+
+	/* cannot go top prev */
+	if (node == NULL)
+		return;
+
+	/* can go top prev */
+	win->top_index--;
+	if (win->ops->needs_blank(win, node, win->top)) // (win, prev_node, curr_node)
+		win->top_index--;
+
+	/* update win->top */
+	win->top = node;
+
+	/* update win->curr - out of cursor line case, cursor must follow top in ranges */
+	while (win->curr_index - win->top_index >= LINES - 2) {
+		/* find prev node */
+		node = win->ops->prev(win, win->curr, false);
+		win->curr_index--;
+
+		/* blank check */
+		if (win->ops->needs_blank(win, node, win->curr))
+			win->curr_index--;
+
+		/* update win->curr */
+		win->curr = node;
+	}
+}
+
+static void tui_window_move_halfpage_up(struct tui_window *win)
+{
+	void *node;
+	int move_cnt = 0;
+
+	/* move win->top */
+	while (move_cnt < (LINES - 2) / 2 - 1) {
```

Comment:
`(LINES - 2) / 2 - 1` can be assigned to a meaningful name at the top of this function.